Objects/listobject.c

```diff
@@ -2584,9 +2572,8 @@ list_index_impl(PyListObject *self, PyObject *value, Py_ssize_t start,
     }
     if (stop < 0) {
         stop += Py_SIZE(self);
-        if (stop < 0)
-            stop = 0;
```

Comment:
I'm not 100% confident that this change is correct. If you consider that stop can no longer be negative, please remove the code an assertion: ``assert(stop >= 0);``.

Reply:
Stop can still be negative at this point. But if `stop <0` then since `start >=0` the for loop is not evaluated. The same optimization is used in `tuple_index_impl`.

One more optimization in `tuple_index_impl` is to take the `i < Py_SIZE(self)` check out of the loop, but we cannot do that here since the length of the list can change (whereas the tuple is immutable)